processor/src/stack/overflow.rs

```diff
@@ -0,0 +1,119 @@
+use super::Felt;
+
+// OVERFLOW
+
+/// OverflowState stores all the operations performed to the
+/// overflow table on a particular clock cycle.
+#[derive(Clone)]
+struct OverflowState {
+    clk: usize,
+    operations: Vec<Option<Felt>>,
+}
+
+/// Overflow table stores a list of overflow values from the stack.
+/// We create a separate structure to be able to store historical changes
+/// in the overflow table, so we can be able to reconstruct what's in the overflow table
+/// at any given point in time.
+#[derive(Clone)]
+pub struct OverflowTable {
+    active: Vec<Felt>,
+    trace: Vec<OverflowState>,
+    trace_enabled: bool,
+}
+
+impl OverflowTable {
+    pub fn new(trace_enabled: bool) -> Self {
+        Self {
+            active: Vec::new(),
+            trace: Vec::new(),
+            trace_enabled,
+        }
+    }
+
+    fn get_state_at(&mut self, step: usize) -> &mut OverflowState {
+        let add_state = match self.trace.last() {
+            None => true,
+            Some(s) if s.clk != step => true,
+            _ => false,
+        };
+
+        if add_state {
+            self.trace.push(OverflowState {
+                clk: step,
+                operations: Vec::new(),
+            });
+        }
+
+        self.trace.last_mut().unwrap()
+    }
+
+    pub fn push(&mut self, step: usize, value: Felt) {
+        self.active.push(value);
+        if self.trace_enabled {
+            self.get_state_at(step).operations.push(Some(value));
+        }
+    }
+
+    pub fn pop(&mut self, step: usize) -> Option<Felt> {
+        let popped = self.active.pop();
+        if popped.is_some() && self.trace_enabled {
+            self.get_state_at(step).operations.push(None);
+        }
+
+        popped
+    }
+
+    /// Get values up to a certain step in the overflow table.
+    pub fn get_values_at(&self, step: usize) -> Vec<Felt> {
+        let mut values: Vec<Felt> = Vec::new();
+        for state in self.trace.iter() {
+            if state.clk > step {
+                break;
+            }
+
+            for op in state.operations.iter() {
+                if let Some(v) = op {
+                    values.push(*v);
+                } else {
+                    values.pop();
+                }
+            }
+        }
+
+        values
```

Comment:
If I understood this correctly, this basically replays all pushes and pops from the beginning to arrive at the current state. This might get expensive for cycles which are far away from the start. Maybe we can do something better?

For example, we could have a binary tree where the key is the cycle and the value is the state of the overflow table. But we'd skip the cycles on which the overflow table doesn't change. So, to find the state at a given cycle you'd basically need to find the entry in the tree which is equal or less than the desired clock cycle. This would result in more memory consumption, but I think the tradeoff might be worth it.

Reply:
I see, I think you had a comment earlier that hopefully we only need to store what's changed, but certainly as cycles grow large a lot more work is needed.
Ok I'll switch it so we basically store versions of the overflow table.